wgpu-core/src/device/life.rs

```diff
@@ -548,21 +548,27 @@ impl<B: GfxBackend> LifetimeTracker<B> {
             return Vec::new();
         }
         let (mut buffer_guard, _) = B::hub(global).buffers.write(token);
-        self.ready_to_map
-            .drain(..)
-            .map(|buffer_id| {
-                let buffer = &mut buffer_guard[buffer_id];
-                let mapping = buffer.pending_mapping.take().unwrap();
-                let result = match mapping.op {
-                    resource::BufferMapOperation::Read { .. } => {
-                        super::map_buffer(raw, buffer, mapping.sub_range, super::HostMap::Read)
-                    }
-                    resource::BufferMapOperation::Write { .. } => {
-                        super::map_buffer(raw, buffer, mapping.sub_range, super::HostMap::Write)
-                    }
-                };
-                (mapping.op, result)
-            })
-            .collect()
+        let mut pending_callbacks: Vec<super::BufferMapPendingCallback> = vec![];
+        for buffer_id in self.ready_to_map.drain(..) {
+            let buffer = &mut buffer_guard[buffer_id];
+            if buffer.life_guard.ref_count.is_none() {
```

Comment:
I said this was the check earlier, but now I see that I missed one step. Let me correct myself here.
The lifetime is split into the following steps:
  1. resource is alive
  2. resource gets dropped by the user - this is where `ref_count == None`, but resource can still be referenced by other things, like bind groups, so we can't totally free it just yet!
  3. typically it's added to `suspected_resources`, which then calls to `remove_abandoned(id)` and stops if this returns false
  4. if it's true, then we add it to free_resources and remove the ID

So what you need to do in this branch is:
```rust
if buffer.life_guard.ref_count.is_none() && self.trackers.buffers.remove_abandoned(id) {
                    hub.buffers.free_id(id);
                    let res = buffer_guard.remove(id).unwrap();
                    self.free_resources.push((res.raw, res.memory)); 
} else { ... }
```